src/event-manager.js

```diff
@@ -67,49 +67,53 @@ class CapturedHandlerEntry {
   }
 }
 
-function handleDelegatedEvent(event) {
-  event.propagationStopped = false;
-  let target = findOriginalEventTarget(event);
+class DelegateHandlerEntry {
+  constructor(eventName, eventManager) {
+    this.eventName = eventName;
+    this.count = 0;
+    this.eventManager = eventManager;
+  }
 
-  while (target && !event.propagationStopped) {
-    if (target.delegatedCallbacks) {
-      let callback = target.delegatedCallbacks[event.type];
-      if (callback) {
-        if (event.stopPropagation !== stopPropagation) {
-          event.standardStopPropagation = event.stopPropagation;
-          event.stopPropagation = stopPropagation;
-        }
-        if ('handleEvent' in callback) {
-          callback.handleEvent(event);
-        } else {
-          callback(event);
+  handleDelegatedEvent(event) {
+    event.propagationStopped = false;
+    let target = findOriginalEventTarget(event);
+
+    while (target && !event.propagationStopped) {
+      if (target.delegatedCallbacks) {
+        let callback = target.delegatedCallbacks[event.type];
+        if (callback) {
+          if (event.stopPropagation !== stopPropagation) {
+            event.standardStopPropagation = event.stopPropagation;
+            event.stopPropagation = stopPropagation;
+          }
+          if ('handleEvent' in callback) {
+            callback.handleEvent(event);
+          } else {
+            callback(event);
+          }
         }
       }
-    }
 
-    target = target.parentNode;
-  }
-}
+      const parent = target.parentNode;
+      const shouldEscapeShadowRoot = this.eventManager.escapeShadowRoot && parent && parent instanceof ShadowRoot;
```

Comment:
```suggestion
      const shouldEscapeShadowRoot = this.eventManager.escapeShadowRoot && parent && parent instanceof ShadowRoot;
      const shouldEscapeShadowRoot = this.eventManager.escapeShadowRoot && parent instanceof ShadowRoot;
```
Would this be a `ReferenceError` in old browser? Sorry I don't remember what we decided to go with this. Or the argument is that if user turns on `escapeShadowRoot`, they should be already aware of potential error here?

Reply:
Using shadow dom is an explicit choice of the user which would make him/her aware of limitations in old browsers so you would only turn this on when supported otherwise escape would be false by default which would make the code never reach the instanceof check.